proposals/cdn-support.md

```diff
@@ -0,0 +1,114 @@
+- Works with static and server output.
+- Works with existing Astro features, like content collections, `astro:assets` images, and `@astrojs/image`.
+
+# Non-Goals
+
+- Auto-prefixing CDN URLs to user-created `<link href>` or `<img src>` etc.
+- Allow changing the CDN URL in runtime.
+- Handling CDN URLs for files in the `public` directory.
+  - Users have to preprend the URL themselves if needed.
+- The Astro CDN service.
+
+# Detailed Design
+
+CDN support is added through the `build.assetsPrefix` config. For all generated asset links in the Astro codebase, before we only consider `base`:
+
+```js
+const assetLink = base + href
+```
+
+Now we also need to consider `assetsPrefix`:
+
+```js
+const assetLink = (assetsPrefix || base) + href
+```
+
+This needs to be applied everywhere, and maybe a utility might help with handling this.
+
+`import.meta.env.ASSETS_PREFIX` is also added for end-users to manually concat strings for assets not controlled by Astro.
+
+Other notes:
+
+1. `assetsPrefix` takes precedence over `base` because `base` applies to the user-facing site only, not the CDN domain:
+  - For example, the user visits `https://example.com/kaboom/` and it fetches assets from `https://cdn.example.com/_astro/explosion.123456.png`
+2. `assetsPrefix` is _the_ prefix for the [`assets` option](https://docs.astro.build/en/reference/configuration-reference/#buildassets), which is `_astro` by default.
+  ```
+  https://cdn.example.com/_astro/explosion.123456.png
+  |---------------------||-----||-------------------|
+       assetsPrefix      assets      assets href
+  ```
+
+# Testing Strategy
+
+We should have these test cases:
+
+- CSS `<link>`
+- Hydration scripts in `<astro-island>`
+- Image `src` with both experimental assets feature and `@astrojs/image`
+- Markdown asset links
+- Content collections asset links
+
+Make sure all links are prefixed with `assetsPrefix` in both static and server builds. A simple integration test to build and check the generated/rendered HTML should be enough.
+
+# Drawbacks
+
+1. This touches a lot of code everywhere.
+2. We need to be concious of `assetsPrefix` if `base` had not already caused us a lot of issues before 🥲
+3. Could be confusing with the `build.assets` option.
+  - "Why is there both `assets` and `assetsPrefix` options?"
+
+# Alternatives
+
+No CDN support. That would block sites that need it.
+
+# Adoption strategy
+
+This should be transparent to all users. Only when `build.assetsPrefix` is used where the feature kicks in.
+
+For third-party Astro libraries that contribute assets externally, they would need to consider `build.assetsPrefix` too. But I'm not aware of any libraries that does this (that's outside of Astro's asset pipeline).
+
+# Unresolved Questions
+
+1. Is the `build.assetsPrefix` name fine? I picked it to complement the existing `build.assets` option. `build.cdnUrl` feels "sticking out".
```

Comment:
That feels fine to me. A bit odd that it's `assetsPrefix` instead of Next's `assetPrefix` but I get that matching `build.assets` is probably more important.